File: Python/trex_selector_methods/trex_da/demo_trex_da_02_equi/demo_trex_da_02_equi.py
# ==============================================================================
# demo_trex_da_02_equi.py
# ==============================================================================
#
# DA-TRex demo and MC simulation for the full-matrix equicorrelated DGP.
#
# Reuses the two-level factor-model DGP (dgp_bt_snr) with rho_within ==
# rho_between, which collapses to pure equicorrelation (compound symmetry).
# The selector uses the EQUI dependency-aware method.
#
#   Part 1: single-run demo.
#   Part 2: MC SNR sweep (rho = 0.25).
#   Part 3: MC rho sweep (SNR = 2.0, rho_within == rho_between).
#
# Mirrors R/trex_selector_methods/trex_da/demo_trex_da_02_equi.R and the equi
# parts of cpp/.../demo_trex_da_02_mc_sim_equi_and_bt. Heavy MC parts are opt-in
# via the RUN_PART_* flags (Part 1 runs by default).
# ==============================================================================

import sys
import os

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

from dgp_generators import dgp_bt_snr
from da_sim_common import (make_support_random, build_da_selector, run_mc,
                           print_table, compute_fdp, compute_tpp)

RUN_PART_1 = True
RUN_PART_2 = False
RUN_PART_3 = False

MC = dict(n=300, p=1000, s=10, amplitude=3.0, n_blocks=10, tFDR=0.2,
          K=20, num_MC=200, seed=2026)


def _run_equi_mc(rho, snr):
    return run_mc(MC["n"], MC["p"], None, MC["amplitude"], snr, MC["tFDR"],
                  MC["K"], MC["num_MC"], MC["seed"], "dgp_bt_snr",
                  {"n_blocks": MC["n_blocks"], "rho_within": rho, "rho_between": rho},
                  "EQUI", {}, s=MC["s"])


# ==============================================================================
# Part 1: single-run demo
# ==============================================================================

def part_1_single_run():
    print("\n" + "=" * 70)
    print("EQUI Demo  |  Part 1: Single-run")
    print("n=150  p=500  s=5  rho=0.25  SNR=2.0")
    print("=" * 70 + "\n")

    support = make_support_random(5, 500, 2026)
    dat = dgp_bt_snr(150, 500, support, amplitude=3.0, n_blocks=5,
                     rho_within=0.25, rho_between=0.25, snr=2.0, seed=2026)
    sel = build_da_selector("EQUI", dat["X"], dat["y"], 0.2, 20, "TLARS")

    selected = sorted(list(sel.selected_indices))
    tsup = dat["true_support"].tolist()
    n_tp = len(set(selected) & set(tsup))

    print("-" * 70)
    print(f"  True support (0-based): {{{', '.join(map(str, tsup))}}}")
    print(f"  Selected: {len(selected)}  |  TP = {n_tp}  FP = {len(selected) - n_tp}")
    print(f"  TPP = {compute_tpp(selected, tsup):.3f}  |  "
          f"FDP = {compute_fdp(selected, tsup):.3f}  (target tFDR <= 0.20)")
    print("-" * 70 + "\n")


# ==============================================================================
# Part 2: MC SNR sweep (rho = 0.25)
# ==============================================================================

def part_2_snr_sweep():
    rho = 0.25
    snr_grid = [0.1, 0.2, 0.5, 1.0, 2.0, 5.0]
    print("=" * 70)
    print(f"EQUI MC — SNR sweep  |  rho={rho:.2f}  {MC['num_MC']} MC")
    print("=" * 70 + "\n")
    rows = []
    for snr in snr_grid:
        print(f"  [SNR = {snr:.2f}] running {MC['num_MC']} MC trials ...")
        rows.append(dict(SNR=snr, **_run_equi_mc(rho, snr)))
    print("\n  Results — SNR sweep:")
    print_table(rows, "SNR", value_fmt="{:<8.2f}", width=8)


# ==============================================================================
# Part 3: MC rho sweep (SNR = 2.0, rho_within == rho_between)
# ==============================================================================

def part_3_rho_sweep():
    snr = 2.0
    rho_grid = [round(0.1 * i, 1) for i in range(10)]
    print("=" * 70)
    print(f"EQUI MC — 1D Pure Equi Sweep  |  SNR={snr:.1f}  {MC['num_MC']} MC")
    print("  Sweeping rho (rho_within == rho_between)")
    print("=" * 70 + "\n")
    rows = []
    for rho in rho_grid:
        print(f"  [rho = {rho:.1f}] running {MC['num_MC']} MC trials ...")
        rows.append(dict(rho=rho, **_run_equi_mc(rho, snr)))
    print("\n  Results — 1D Pure Equi Sweep:")
    print_table(rows, "rho", value_fmt="{:<8.2f}", width=8)


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    if RUN_PART_1:
        part_1_single_run()
    if RUN_PART_2:
        part_2_snr_sweep()
    if RUN_PART_3:
        part_3_rho_sweep()
    print("\ndemo_trex_da_02_equi.py complete.")

File: R/trex_selector_methods/trex_da/demo_trex_da_02_equi/demo_trex_da_02_equi.R
# ==============================================================================
# demo_trex_da_02_equi.R
# ==============================================================================
# DA-TRex demo and MC simulation for the full-matrix equi-correlated DGP.
#
# Reuses the BT factor-model DGP by setting rho_within == rho_between.
# ==============================================================================

library(TRexSelectorNeo)
library(plotly)
library(parallel)

num_cores <- 6L

this_dir_ <- tryCatch(dirname(normalizePath(sys.frame(1)$ofile)),
                      error = function(e) {
                        args <- commandArgs(trailingOnly = FALSE)
                        file_arg <- grep("--file=", args, value = TRUE)
                        if (length(file_arg) > 0) {
                          dirname(normalizePath(sub("--file=", "", file_arg[1])))
                        } else {
                          "."
                        }
                      })

# for make_support_random
source(file.path(this_dir_, "..", "..", "support_generators.R"))

# for dgp_bt_snr
source(file.path(this_dir_, "..", "trex_da_dgps.R"))
source(file.path(this_dir_, "..", "..", "simulation_utils.R"))

# ==============================================================================
# Part 1: Single-run demo
# n=150, p=500, s=5, rho=0.25, SNR=2.0
# ==============================================================================
if (FALSE) {

  trx_da_02_equi_01 <- function() {
    cat("\n", strrep("=", 70), "\n", sep = "")
    cat("  EQUI Demo | Part 1: Single-run\n")
    cat("  n=150 p=500 s=5 rho=0.25 SNR=2.0\n")
    cat(strrep("=", 70), "\n\n")

    support_p1 <- make_support_random(s = 5L, p = 500L, seed = 2026L)

    # Set rho_within and rho_between both to 0.25 to make it pure equi-correlated
    dat_p1 <- dgp_bt_snr(
      n           = 150L,
      p           = 500L,
      support     = support_p1,
      amplitude   = 3.0,
      n_blocks    = 5L,
      rho_within  = 0.25,
      rho_between = 0.25,
      snr         = 2.0,
      seed        = 2026L
    )

    plot_cormat(cor(dat_p1$X))

    res_p1 <- TRexSelectorNeo::TRexDASelector$new(
      dat_p1$X, dat_p1$y, tFDR = 0.2, seed = -1L, verbose = FALSE,
      da_control = TRexSelectorNeo::trex_da_control(da_method = "EQUI"),
      control = TRexSelectorNeo::trex_control(solver = "TLARS", K = 20L))
    res_p1$select()

    cat(sprintf("  Selected variables: %d\n\n", sum(res_p1$selected_var)))

  }
  # Run demo for Part 1
  trx_da_02_equi_01()

} # end of part 1
# ==============================================================================


# ==============================================================================
# Part 2: Monte Carlo simulation — SNR sweep
# Fixed rho_within = 0.25, rho_between = 0.25.
# Swept: SNR in {0.1, 0.2, 0.5, 1.0, 2.0, 5.0}
# ==============================================================================
if (TRUE) {

  trx_da_02_equi_02 <- function() {
    MC_PARAMS <- list(
      n           = 300L,
      p           = 1000L,
      s           = 10L,
      amplitude   = 3.0,
      rho_within  = 0.25,
      rho_between = 0.25,
      tFDR        = 0.2,
      K           = 20L,
      num_MC      = 200L,
      seed        = 2026L
    )

    snr_grid <- c(0.1, 0.2, 0.5, 1.0, 2.0, 5.0)

    cat(strrep("=", 70), "\n")
    cat(sprintf(" EQUI MC — SNR sweep | rho=%.2f, %d MC\n",
                MC_PARAMS$rho_within, MC_PARAMS$num_MC))
    cat(strrep("=", 70), "\n\n")

    snr_results <- lapply(snr_grid, function(snr_val) {
      cat(sprintf("\n [SNR = %.2f] running %d MC trials ...\n", snr_val, MC_PARAMS$num_MC))
      r <- .run_mc_equi(
        n        = MC_PARAMS$n,
        p        = MC_PARAMS$p,
        support  = NULL,
        amplitude = MC_PARAMS$amplitude,
        s        = MC_PARAMS$s,
        rho      = MC_PARAMS$rho_within,
        n_blocks = 10L,
        snr      = snr_val,
        tFDR     = MC_PARAMS$tFDR,
        K        = MC_PARAMS$K,
        num_MC   = MC_PARAMS$num_MC,
        seed     = MC_PARAMS$seed,
        n_cores  = num_cores
      )
      c(list(SNR = snr_val), r)
    })

    cat("\n\n Results — SNR sweep:\n")
    .print_table(snr_results, "SNR", "%-8.2f")
  }

  # Run demo for Part 2
  trx_da_02_equi_02()

} # end of part 2
# ==============================================================================


# ==============================================================================
# Part 3: Monte Carlo simulation — 1D Pure Equi Sweep
# Fixed SNR = 2.0.
# Swept: rho in seq(0.0, 0.9, by = 0.1).
# ==============================================================================
if (TRUE) {

  trx_da_02_equi_03 <- function() {
    MC_PARAMS_P3 <- list(
      n         = 300L,
      p         = 1000L,
      s         = 10L,
      amplitude = 3.0,
      snr       = 2.0,
      tFDR      = 0.2,
      K         = 20L,
      num_MC    = 200L,
      seed      = 2026L,
      n_blocks  = 10L
    )

    rho_grid <- seq(0.0, 0.9, by = 0.1)

    cat(strrep("=", 70), "\n")
    cat(sprintf(" EQUI MC — 1D Pure Equi Sweep | SNR=%.1f, %d MC\n",
                MC_PARAMS_P3$snr, MC_PARAMS_P3$num_MC))
    cat(" Sweeping rho (where rho_within == rho_between)\n")
    cat(strrep("=", 70), "\n\n")

    rho_results <- lapply(rho_grid, function(rho_val) {
      cat(sprintf("\n [rho = %.1f] running %d MC trials ...\n", rho_val, MC_PARAMS_P3$num_MC))
      r <- .run_mc_equi(
        n        = MC_PARAMS_P3$n,
        p        = MC_PARAMS_P3$p,
        support  = NULL,
        amplitude = MC_PARAMS_P3$amplitude,
        s        = MC_PARAMS_P3$s,
        rho      = rho_val,
        n_blocks = MC_PARAMS_P3$n_blocks,
        snr      = MC_PARAMS_P3$snr,
        tFDR     = MC_PARAMS_P3$tFDR,
        K        = MC_PARAMS_P3$K,
        num_MC   = MC_PARAMS_P3$num_MC,
        seed     = MC_PARAMS_P3$seed,
        n_cores  = num_cores
      )
      c(list(rho = rho_val), r)
    })

    cat("\n\n Results — 1D Pure Equi Sweep:\n")
    .print_table(rho_results, "rho", "%-8.2f")
  }

  # Run demo for Part 3
  trx_da_02_equi_03()

} # end of part 3
# ==============================================================================
